### DISCUSSION TITLE
Can I use Eval annex for completions?

### DISCUSSION THREAD

**Discussion by anihm136** [z-shell/zi#268] [Q&A] [answered]
I just came across the eval annex, and am able to use it for slow initialization commands as described in the documentation. I also have a few commands which generate a completion file, which needs to be placed in fpath. Can I use the eval annex to generate these completions and add them to fpath on update? Or is there a better way to do this (as compared to manually generating the completion and pointing to it as a snippet)? Thanks

**Comment by ss-o** [comment, marked as answer]
Hi @anihm136 :wave: 


> Can I use the eval annex to generate these completions...


Technically, you can, but I don't think this is what you want, as the idea of the `eval` annex is to speed up shell startup time because of the slow commands initialization by providing a workaround, however, in your case, it seems that your "few commands which generate a completion" **is not** and **should not** be executed at startup, because the completions can be prepared at any point (before/after startup).


Correct me if I misunderstood something :+1: 


> which needs to be placed in fpath.


This is done automatically ([zsh_plugin_standard#functions-directory](https://wiki.zshell.dev/community/zsh_plugin_standard#functions-directory)) or manually passing the completion control for **Zi** (1.[overview#the-completion-management](https://wiki.zshell.dev/docs/getting_started/overview#the-completion-management), 2.[commands#completions-management](https://wiki.zshell.dev/docs/guides/commands#completions-management)). 🚀


> Can I use the eval annex to generate these completions and add them to fpath on update.


There are many ways to run [command execution after cloning, updating, or loading](https://wiki.zshell.dev/docs/guides/syntax/ice-modifiers#command-execution-after-cloning-updating-or-loading), so `eval` usage would be redundant 🛑 ✋


> Or is there a better way to do this (as compared to manually generating the completion and pointing to it as a snippet)? 


To be able to make a suggestion I need to know the exact steps you are trying to accomplish. If could provide step-by-step instructions, then I would be able to help sort this out 🪄 💡

**Reply by anihm136** [reply]
Thanks for the reply. Let me clarify what I'm trying to achieve -
- For some command, let's say the completion file is generated using the command `cmd --completion zsh`
- The output of this command should be sent to a file, say `_cmd`
- This file should then be placed somewhere on fpath
- Finally, it should be periodically updated (for example, if the tool is updated and new completions come in)

What I'm trying to achieve here is to remove the manual steps of piping to a file and managing that file myself, and update the completions whenever I run `zi update` - how can I set this up?

**Reply by anihm136** [reply]
The reason I considered eval annex is because here also, the command produces output that can be cached. However, instead of sourcing this output, I would like to place the output on fpath

**Reply by ss-o** [reply]
> For some command, let's say the completion file is generated using the command `cmd --completion zsh`

So you want to generate a completion at the plugin install time and load the completion, then repeat the same process at update?

Can you name the specific plugin or provide an install snippet? Or do you want to create your own plugin which handles some commands for you?

> I would like to place the output on fpath

The output is automatically added to fpath by the plugin unless it is not a plugin, then hook/wrapper may be required.


Without a specific example, I am unable to provide an answer as I am unaware of what are trying to install.

**Reply by anihm136** [reply]
This isn't tied to a plugin specifically - my usecase is more similar to sourcing a snippet as a completion. For example -
```sh
zi ice as"completion"
zi snippet <path/to/file>
```

Here, instead of `<path/to/file>`, I would like to provide a command which generates the contents of the file

**Reply by ss-o** [reply]
It doesn't make much sense or I am missing out on something 💁‍♂️

I can provide the steps to execute the command with output added to the `$fpath` and repeated on the update.

However, what you are trying to do is to generate your own completions, so it would make more sense to do the following:

- create a repository that uses scheduled GitHub actions to fetch/generate/update all required completions.
- setup the repository as a zsh plugin.
- issue command: `zi light anihm136/zsh-comps`

This will allow managing personal completions which can be extended at any time, are automated, and are easy to set up :dart: 

Let me know which method is the preferred one.

**Reply by anihm136** [reply]
> create a repository that uses scheduled GitHub actions to fetch/generate/update all required completions.

That option had not occurred to me, thanks for suggesting it! I will try this out and see if it is suitable - I should be able to work it out. In the meantime, could you also share an example of the first approach?

**Reply by ss-o** [reply]
It can be done in multiple ways using appropriate [ice modifiers](https://wiki.zshell.dev/docs/guides/syntax/ice-modifiers). Based on your description:

```sh
_gen_cmd_comp() {
  # Your function which generates completion, e.g.
  echo zsh > _cmd
}

zi ice id-as'comp/name' as'completion' nocompile \
  atclone'_gen_cmd_comp' atpull'%atclone' run-atpull
zi light z-shell/0
```

Interacting with completion:

```sh
zi clist | grep 'comp/name'
```
```sh
zi report comp/name
```
```sh
zi cdisable _cmd
```
```sh
zi cenable _cmd
```

Completion multiplier?

```sh
_gen_cmd_comp() {
  # Your function which generates multiple completions, e.g.
  echo zsh1 > _cmd1
  echo zsh2 > _cmd2
  echo zsh3 > _cmd3
  echo zsh4 > _cmd4
  echo zsh5 > _cmd5
  echo zsh6 > _cmd6
  echo zsh7 > _cmd7
  echo zsh8 > _cmd8
  echo zsh9 > _cmd9
  echo zsh10 > _cmd10
}

zi ice id-as'comp/name' as'completion' nocompile \
  atclone'_gen_cmd_comp' atpull'%atclone' run-atpull
zi light z-shell/0
```

```sh
zi report comp/name
```

Generate completions on the go?

```sh
# Set parameters
GENCOMPL_FPATH="${ZI[SNIPPETS_DIR]}/comps"
GENCOMP_DIR="${ZI[SNIPPETS_DIR]}/comps"

# Install annex
zi light-mode for \
  z-shell/z-a-submods

# Go?
zi ice wait trigger-load'!gcomp' blockf \
  atclone'command rm -rf lib/*;git ls-files -z lib/ | xargs -0 git update-index --skip-worktree' \
  submods'RobSis/zsh-completion-generator -> lib/zsh-completion-generator; 
    nevesnunes/sh-manpage-completions -> lib/sh-manpage-completions' \
  atload'gcomp(){gencomp "${@}" && zi creinstall -q ${ZI[SNIPPETS_DIR]}/comps 1>/dev/null}'
zi light Aloxaf/gencomp
```

Generate `fzf` completion:

```sh
gcomp fzf
```

> This is unusual as Zi should be able to install plugins/completions, and here creates while installing, so made it simple to understand common ice modifiers.

**Reply by anihm136** [reply]
Got it! Will try out these two approaches and see which one is easiest to work with for me. Thank you so much for your patient explanation, it was really helpful!

**Reply by ss-o** [reply]
If you decide to use a zsh plugin approach, then you can create and provide it as a zsh plugin/package. This method would allow choosing completion profiles:

Install default profile:

```sh
zi for comps-name
```

Install macOS developer completion profile:

```sh
zi pack'mac-dev' for comps-name
```

To achieve this you need to set [ZI[PKG_OWNER]](https://wiki.zshell.dev/docs/guides/customization#modify-settings) or create repository under @z-shell, then include `package.json` with profiles install commands:

```json
    "zi-ices": {
      "default": {
        "git": "",
        "lucid": "",
        "as": "completion",
        "atpull": "%atclone",
        "nocompile": ""
      },
      "mac-dev": {
        "git": "",
        "lucid": "",
        "as": "completion",
        "atpull": "%atclone",
        "nocompile": ""
      }
    }
```

> This is just an example. The `package.json` includes ice modifiers as it would be in `.zshrc` just using JSON syntax.
